generators/server/templates/src/main/java/package/security/oauth2/AuthorizationHeaderUtil.java.ejs

```diff
@@ -18,46 +18,162 @@
+        if (authentication instanceof OAuth2AuthenticationToken) {
+            OAuth2AuthenticationToken oauthToken = (OAuth2AuthenticationToken) authentication;
+            String name =  oauthToken.getName();
+            String registrationId = oauthToken.getAuthorizedClientRegistrationId();
             OAuth2AuthorizedClient client = clientService.loadAuthorizedClient(
-                oauthToken.getAuthorizedClientRegistrationId(),
-                oauthToken.getName());
-            accessToken = client.getAccessToken().getTokenValue();
-        } else if (authToken instanceof JwtAuthenticationToken) {
-            accessToken = ((JwtAuthenticationToken) authToken).getToken().getTokenValue();
-        }
+            registrationId,
+            name);
+
+            if (null == client) {
+                throw new OAuth2AuthorizationException(new OAuth2Error("access_denied", "The token is expired", null));
+            }
+            OAuth2AccessToken accessToken = client.getAccessToken();
+
+            if (accessToken != null) {
+                String tokenType = accessToken.getTokenType().getValue();
+                String accessTokenValue = accessToken.getTokenValue();
+                if (isExpired(accessToken)) {
+                    log.info("AccessToken expired, refreshing automatically");
+                    accessTokenValue = refreshToken(client, oauthToken);
+                    if (null == accessTokenValue) {
+                        SecurityContextHolder.getContext().setAuthentication(null);
+                        throw new OAuth2AuthorizationException(new OAuth2Error("access_denied", "The token is expired", null));
+                    }
+                }
+                String authorizationHeaderValue = String.format("%s %s", tokenType, accessTokenValue);
+                return Optional.of(authorizationHeaderValue);
+            }
 
-        if (accessToken == null) {
-            return Optional.empty();
-        } else {
-            String tokenType = "Bearer";
-            String authorizationHeaderValue = String.format("%s %s", tokenType, accessToken);
+        } else if (authentication instanceof JwtAuthenticationToken) {
+            JwtAuthenticationToken accessToken = (JwtAuthenticationToken) authentication;
+            String tokenType = (String) accessToken.getToken().getClaims().get("typ");
+            String tokenValue = accessToken.getToken().getTokenValue();
+            String authorizationHeaderValue = String.format("%s %s", tokenType, tokenValue);
             return Optional.of(authorizationHeaderValue);
         }
+        return Optional.empty();
+    }
+
+    private String refreshToken(OAuth2AuthorizedClient client, OAuth2AuthenticationToken oauthToken) {
+        OAuth2AccessTokenResponse atr = refreshTokenClient(client);
+        if (atr == null || atr.getAccessToken() == null) {
+            log.info("Failed to refresh token for ${currentUser.name}");
```

Comment:
not sure if this will work ` ${currentUser.name}`?

Reply:
Well identified! Done